### PR TITLE
add function to swap connected nodes in DAGCircuit

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
Sometimes when doing circuit optimization it is useful to be able to swap two connected gates for instance when they commute. If the two gates are on the same qubits this can be done with the existing method `substitute_node`, however when the qubits are disjoint this can't be used.  This pr adds a method `swap_nodes` which allows this operation.

### CODE REVIEW

File: qiskit/dagcircuit/dagcircuit.py [hunk 1]
```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
...
+        Raises:
+            DAGCircuitError: if either node is not an OpNode
+        """
+        if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
+            raise DAGCircuitError("nodes to swap are not both OpNodes")
+        try:
+            connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
+        except rx.NoEdgeBetweenNodes:
+            # since nodes aren't connected there's nothing to do
+            return
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1390]: This feels like it ought to be some sort of exception - we're being asked to swap nodes that can't be swapped.  I don't think silently doing nothing is safe.

File: qiskit/dagcircuit/dagcircuit.py [hunk 2]
```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
...
+
+        Args:
+            node1 (OpNode): predecessor node
+            node2 (OpNode): successor node
+
+        Raises:
+            DAGCircuitError: if either node is not an OpNode
+        """
+        if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
+            raise DAGCircuitError("nodes to swap are not both OpNodes")
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1385]:
```suggestion
            raise DAGCircuitError("nodes to swap are not both DAGOpNodes")
```
(so the user can more easily find the correct type, if they see this error)

File: qiskit/dagcircuit/dagcircuit.py [hunk 3]
```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
...
+        try:
+            connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
+        except rx.NoEdgeBetweenNodes:
+            # since nodes aren't connected there's nothing to do
+            return
+        node1_id = node1._node_id
+        node2_id = node2._node_id
+        edge_find = lambda x: x == edge
+        for edge in connected_edges[::-1]:
+            edge_parent = self._multi_graph.find_predecessors_by_edge(node1_id, edge_find)[0]
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1395]: The delayed closure of this `lambda` is _very_ confusing.  At the time the `lambda` is written, the variable `edge` would return `UnboundLocalException` on access.  It will only succeed later, within the loop, once `edge` has been assigned to for the first time.

I'm surprised pylint isn't complaining quite a lot about this for that reason, and because it usually doesn't like direct assignment of lambdas.

File: qiskit/dagcircuit/dagcircuit.py [hunk 4]
```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
...
+            self._multi_graph.remove_edge(edge_parent._node_id, node1_id)
+            self._multi_graph.add_edge(edge_parent._node_id, node2_id, edge)
+            edge_child = self._multi_graph.find_successors_by_edge(node2_id, lambda x: x == edge)[0]
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1398]: I don't know if there's an equivalent for `predecessors`, but this method could be `find_adjacent_node_by_edge` to avoid calculating the whole list.

File: qiskit/extensions/quantum_initializer/uc.py [hunk 5]
```diff
@@ -135,7 +135,7 @@ def _dec_ucg(self):
         the diagonal gate is also returned.
         """
         diag = np.ones(2**self.num_qubits).tolist()
-        q = QuantumRegister(self.num_qubits)
+        q = QuantumRegister(self.num_qubits, name="q")
```
Comment (jakelishman) [inline qiskit/extensions/quantum_initializer/uc.py:138]: This seems unrelated?

File: test/python/dagcircuit/test_dagcircuit.py [hunk 6]
```diff
@@ -2131,5 +2132,97 @@ def test_clbit_conditional(self):
         )
 
 
+class TestSwapNodes(QiskitTestCase):
+    """Test Swapping connected nodes."""
+
+    def test_1q_swap_fully_connected(self):
+        """Test swapping single qubit gates"""
+        from qiskit.converters import dag_to_circuit
```
Comment (jakelishman) [inline test/python/dagcircuit/test_dagcircuit.py:2140]: This import could be module-level.

File: test/python/dagcircuit/test_dagcircuit.py [hunk 7]
```diff
@@ -2131,5 +2132,97 @@ def test_clbit_conditional(self):
...
+
+        qreg = QuantumRegister(1)
+        circuit = QuantumCircuit(qreg)
+        circuit.rx(pi, qreg[0])
+        circuit.rx(pi / 2, qreg[0])
+        dag = circuit_to_dag(circuit)
+        op_nodes = dag.op_nodes()
```
Comment (jakelishman) [inline test/python/dagcircuit/test_dagcircuit.py:2147]: This dance might be a little easier if we just built the `DAGCircuit` directly - the return value from `apply_operation_back` is the node index we're inferring here.

File: test/python/dagcircuit/test_dagcircuit.py [hunk 8]
```diff
@@ -2131,5 +2132,97 @@ def test_clbit_conditional(self):
...
+        dag.swap_nodes(op_nodes[0], op_nodes[1])
+        swapped_circuit = dag_to_circuit(dag)
+
+        expected_circuit = QuantumCircuit(4)
+        expected_circuit.cx(1, 0)
+        expected_circuit.cz(2, 3)
+        self.assertEqual(swapped_circuit, expected_circuit)
+
```
Comment (jakelishman) [inline test/python/dagcircuit/test_dagcircuit.py:2225]: These tests seem fairly conclusive for the case of only having the two operations in question in the circuit.  Can we add some tests that have operation successors and predecessors of the two nodes as well, potentially with the successors/predecessors having unrelated edges connecting them to one of the pair?

I imagine it should work, since we're already dealing with the in- and out-nodes, but the multi-wire case would be good.

**Review by jakelishman — COMMENTED** [review]
Do you have a particular case / transpiler pass in mind for when you want this?  There's a general hope (cc: @alexanderivrii) that we might be able to use `DAGDependency` as our principle driver type for dealing with commutation relations, though that's likely to be a long way off.  If this is useful in the near-term, it generally seems like a useful helper.

Don't forget the release note.

File: qiskit/dagcircuit/dagcircuit.py (hunk 1 above)
Comment (ewinston) [reply qiskit/dagcircuit/dagcircuit.py:1390]: The case I was thinking of for not raising, which this doesn't quite cover, is was a situation like,
```
q_0: ──■───────
       │       
q_1: ──┼────■──
     ┌─┴─┐  │  
q_2: ┤ X ├──┼──
     └───┘┌─┴─┐
q_3: ─────┤ X ├
          └───┘
```
where the method might be looping through some list of many swap candidates. When it encounters this one it can merrily  continue without having to do anything. I'm not sure of an easy way, however, to catch when operations can't be considered to be swapped due to other intervening instructions. Maybe `DAGDependency` would work better for such a case. I'll raise here.

File: qiskit/dagcircuit/dagcircuit.py (hunk 3 above)
Comment (ewinston) [reply qiskit/dagcircuit/dagcircuit.py:1395]: I added `edge` as a default argument to the lambda which I understand avoids the delayed closure.

File: qiskit/dagcircuit/dagcircuit.py (hunk 4 above)
Comment (ewinston) [reply qiskit/dagcircuit/dagcircuit.py:1398]: How do you make `find_adjacent_node_by_edge` filter on predecessor or successor?

File: test/python/dagcircuit/test_dagcircuit.py (hunk 8 above)
Comment (ewinston) [reply test/python/dagcircuit/test_dagcircuit.py]: I couple of tests of this nature have been added

File: qiskit/dagcircuit/dagcircuit.py [hunk 9]
```diff
@@ -1828,7 +1857,7 @@ def properties(self):
         }
         return summary
 
-    def draw(self, scale=0.7, filename=None, style="color"):
+    def draw(self, scale=0.7, filename=None, style="color", **kwargs):
```
Comment (jakelishman) [inline qiskit/dagcircuit/dagcircuit.py:1860]: Can we shift this to a different PR? It seems unrelated.

File: qiskit/visualization/dag_visualization.py [hunk 10]
```diff
@@ -25,7 +25,7 @@
 
 
 @_optionals.HAS_GRAPHVIZ.require_in_call
-def dag_drawer(dag, scale=0.7, filename=None, style="color"):
+def dag_drawer(dag, scale=0.7, filename=None, style="color", show_node_id=False):
```
Comment (jakelishman) [inline qiskit/visualization/dag_visualization.py:28]: Similarly here - it seems like an unrelated change.

**Review by jakelishman — COMMENTED** [review]
Thanks for the new tests.  Aside from my wanting to shift the unrelated visualisation-related changes to a different PR, I think this is fine to merge.

File: qiskit/visualization/dag_visualization.py (hunk 10 above)
Comment (ewinston) [reply qiskit/visualization/dag_visualization.py]: removed in ff987f1c6

File: qiskit/dagcircuit/dagcircuit.py (hunk 9 above)
Comment (ewinston) [reply qiskit/dagcircuit/dagcircuit.py]: removed in ff987f1c6

**Review by kdk — APPROVED** [review]
